### PR TITLE
ci(github): replace docker.pkg.github.com with quay.io

### PR DESCRIPTION
**Pull request by DrSensor**
<sup>related pr/issues: #701, #625</sup>
> see [here](https://quay.io/repository/drsensor/nu?tab=tags) for what it looks like
tips: sort list by **Tag** or **Size**

### Demo
```sh
docker run -it quay.io/drsensor/nu:$TAG
```

### Why
GitHub plan to move all access of their package registry using `GITHUB_TOKEN` ([ref1], [ref2], [ref3]). Until the migration reach `docker.pkg.github.com`, I want to change it into quay.io for the next release so I can test some stuff in an isolated environment.

[ref1]: https://github.blog/2019-09-11-proxying-packages-with-github-package-registry-and-other-updates/#using-packages-in-actions
[ref2]: https://github.blog/changelog/2019-09-06-publishing-a-package-no-longer-creates-a-release-and-other-github-package-registry-updates/
[ref3]: https://github.blog/changelog/2019-08-16-publishing-packages-from-a-github-action-to-github-package-registry/

### How to change the DOCKER_REGISTRY token to quay.io
0. login quay.io
1. visit <https://quay.io/user/nushell?tab=robots>
2. click **Create Robot Account** button then name it as `action`
Create Robot Account dialog
3. Change **PERMISSION** to **Write**
change permission dialog
4. Click **View Credentials**
options dialog
5. Copy the **Robot Token**
credential dialog
6. go to <https://github.com/nushell/nushell/settings/secrets>
7. overwrite *DOCKER_REGISTRY* secret with the copied **Robot Token**
github secrets page

### CODE REVIEW

**Comment by jonathandturner** [comment]
I logged in quay.io but the link to https://quay.io/user/nushell?tab=robots doesn't work.

**Comment by DrSensor** [comment]
@jonathandturner are you logged in as an organization?

**Comment by jonathandturner** [comment]
I tried to, but I didn't see a way to do so.

**Comment by DrSensor** [comment]
got it, @jonathandturner here is the correct link <https://quay.io/organization/nushell?tab=robots>

the dialog will look something like this
image

**Comment by jonathandturner** [comment]
If I click on the link I see this:

image

I don't see a "create robot" button.  If I click on one of those links, I see, for example:

image

Still don't see anything that's talking about robots.

**Comment by jonathandturner** [comment]
I wonder if it's possible we haven't configured the organization to show me all the buttons on the left that you have in your screenshot

**Comment by DrSensor** [comment]
cc: @vsoch

**Comment by vsoch** [comment]
Yes, we already having quay.io, and we also have robot tokens that deploy from CircleCi. What exactly is needed here?

**Comment by vsoch** [comment]
Before making any changes, I'd like to discuss the namespace here. This set of containers built via GitHub CI is going to have the same namespace as the CircleCI build, and this conflict is an issue. Specifically, CircleCI builds nushell/nu and nushell/nu-base (see https://quay.io/organization/nushell) and for each, we have the following tags

 - *nushell/nu:devel* development build on merge to master
 - *nushell/nu:nightly* complete (release) build, nightly cron 
 - *nushell/nu:latest* on release, also *nushell:nu:<version>*

Importantly, we don't want the namespace to collide, each CI service should be scoped to a specific function / set of builds. 

Also, we want to be absolutely sure that GitHub Actions / CI is out of beta - I wouldn't store secrets (even robot tokens) there until that is the case.

**Comment by jonathandturner** [comment]
@vsoch / @DrSensor - what are the next steps we need to do for this PR? I'm happy to continue chatting through what needs to happen.  @DrSensor - are you interested in continuing to work on this?

**Comment by DrSensor** [comment]
Absolutely! But I'm not sure what to do next 🤔. Btw, GitHub action and registry will be moving to general availability on November 13.

**Comment by vsoch** [comment]
The namespace of containers needs to be discussed and decided upon.

**Comment by vsoch** [comment]
I think I outlined this clearly in https://github.com/nushell/nushell/pull/806#issuecomment-542259590

**Comment by DrSensor** [comment]
So, if I just exclude the latest tag then it should be fine 🤔

**Comment by vsoch** [comment]
That would work! Since we would be required to add a bot token (I can do this) I think we should be conservative and wait until it's officially out of beta. If something (unlikely) happens, we wouldn't want the fallback to be "But we were in beta, you were warned." 

In the meantime, I'd like to request a clean table of image uris and a brief description added to the README of the Docker folder. For example, if I go there I should immediately see all the images available, and understand enough about the tags to know which one is optimal for my use case.

**Comment by DrSensor** [comment]
Is `mv docs/docker.md docker/README.md` will do?

**Comment by vsoch** [comment]
Didn't even see that! Beautifully done. I'd just add the current quay.io images there, and also specifically mention that the registry is quay.io (folks will use uri `nu:<tag>` and not `quay.io/nu:<tag>`).  For the docker/README.md this serves a different purpose, it helps developers to build the containers on their own. So we don't want to move the file to overwrite it, but rather add a note (and link) to the table, and I would want to have an added section for how the user can build the container tags specified in the GitHub Actions Workflow. Does that make sense?

 - docker/README.md: is how to build any given tag (and link to table)
 - docs/docker.md: describes all tags

**Comment by vsoch** [comment]
Just wanting to double check here - when actions goes out of beta it won't be available (unless it's educational or other discount or enterprise supported). Will the nushell organization still have them?

**Comment by DrSensor** [comment]
I though it's going to be free for OSS project :thinking: 
The significant limitation is only on the cache size, the number of concurrent workflows, and included minutes.

Also, I've noticed that github began removing some of software in the `macOS-latest`, including android and IOS build tools. Rust toolchain is not available for macOS from the beginning.

**Comment by vsoch** [comment]
ah perfecto! A few of my grad school labs are getting emails (and I get them as well being a member of the org) that Actions are running out, and this is logical if they aren't flagged as free for open source. We should be good then :)

Thinking more about scoping, since there are many more images here, and they are versioned, I'm thinking that CircleCI can handle the nightly and devel tags, and then here we can do versioned pushes (based on the regexp for the branch matching a version tag too) and the image names in this file:

https://github.com/nushell/nushell/blob/master/docs/docker.md

And that makes sense because there are many more images here, and it wouldn't make sense to have them frequently build. Next, let's take a look at envars that are needed in secrets (not that I'm not an admin, I can see settings for the repo but not secrets):

```bash
REGISTRY=${REGISTRY,,}; export TAG=${GITHUB_REF##*/}-${{ matrix.tag }}
```
`REGISTRY` would need to be quay.io, correct? `GITHUB_REF` would be the tag or branch (typically a release?) A few quick questions - are we versioning using semver, and are the versions being prefixed with "v" ?

For the below login, it would be good to put the password and user variable in quotes. Could we also be consistent in the namespace for the variables? E.g.:

 - DOCKER_REGISTRY
 - DOCKER_PASSWORD
 - DOCKER_USER

or something along those lines. In case it matters, the format that is suggested on quay.io looks like

```bash
docker login -u="${DOCKER_USER}" -p="${DOCKER_PASSWORD}" ${DOCKER_REGISTRY}
```
as compared to what we have (should be functionally the same)
```bash
echo ${{ secrets.DOCKER_REGISTRY }} | docker login ${REGISTRY%/*} -u ${USER,,} --password-stdin
```

A small comment, I know we are "waiting for" the GitHub package registry, but what would be so terrible about consistently using quay.io, until there is reason to do otherwise?

I wouldn't hard code the USER and REGISTRY based on the GitHub actor, because that may not always be the case, these lines:
```
USER: ${{ github.actor }}+action
REGISTRY: quay.io/${{ github.actor }}
```

but maybe that's not intended for Quay.io.

Anyhoo, once we fix up these details I can create a token - @jonathandturner I don't have access to see secrets so if you want to give me temporary access I can add it, otherwise I can add your login (do you have one?) to quay.io to the nushell organization there.

**Comment by DrSensor** [comment]
replying @vsoch 

> Thinking more about scoping, since there are many more images here, and they are versioned, I'm thinking that CircleCI can handle the nightly and devel tags, and then here we can do versioned pushes (based on the regexp for the branch matching a version tag too) and the image names in this file:

I agree with this, this mean `latest` and all SemVer tags will be published from GithubAction while `nightly` and `devel` tags will be published from CircleCI 🤔

---
> REGISTRY would need to be quay.io, correct? GITHUB_REF would be the tag or branch (typically a release?) A few quick questions - are we versioning using semver, and are the versions being prefixed with "v" ?

Yes,
- `${REGISTRY,,}` value is `quay.io/nuhsell` and
- `GITHUB_REF` value is `refs/tags/v?[0-9]+.[0-9]+.[0-9]+*`
  - which `${GITHUB_REF##*/}` will extract the tag name

So far, nushell git-tag use SemVer without 'v' (see [here](https://github.com/nushell/nushell/tags)).

---
<blockquote>

For the below login, it would be good to put the password and user variable in quotes. Could we also be consistent in the namespace for the variables? E.g.:

    DOCKER_REGISTRY
    DOCKER_PASSWORD
    DOCKER_USER

or something along those lines. In case it matters, the format that is suggested on quay.io looks like
```
docker login -u="${DOCKER_USER}" -p="${DOCKER_PASSWORD}" ${DOCKER_REGISTRY}
```
</blockquote>

I agree with this since the environment variable in gh-action workflows can be defined in the `job` scope. One question though, is it better to use 
```
echo ${DOCKER_PASSWORD} | docker login --password-stdin ...
```
instead of
```
docker login -p="${DOCKER_PASSWORD}" ...
```
???

---
<blockquote>

I wouldn't hard code the USER and REGISTRY based on the GitHub actor, because that may not always be the case, these lines:
```
USER: ${{ github.actor }}+action
REGISTRY: quay.io/${{ github.actor }}
```
but maybe that's not intended for Quay.io.
</blockquote>

Ah yes, I declare it that way to make changes easier. I will replace `${{ github.actor }}` with `nushell`.

**Comment by vsoch** [comment]
@DrSensor your changes look good, and I think we're ready to test with a deployment (and can tweak if necessary). For the docker login bit, I've seen it both ways and I think they are functionally equivalent - yours might be even better since it's from stdin.

@jonathandturner the last bit to do is to add a token to the settings here. I can do this with (temporary) admin, or I can add you to the quay.io/nushell organization, if you can create an account and share your user name.

File: .github/workflows/docker-publish.yml [hunk 1]
```diff
@@ -31,6 +35,10 @@ jobs:
     name: Build and publish docker images
     needs: compile
     runs-on: ubuntu-latest
+    env:
+      DOCKER_REGISTRY: quay.io/nushell
+      DOCKER_PASSWORD: ${{ secrets.DOCKER_REGISTRY }}
+      DOCKER_USER: nushell+action # TBD
```
Comment (vsoch) [inline .github/workflows/docker-publish.yml]: @DrSensor please have this be an environment variable as well.

Comment (vsoch) [reply .github/workflows/docker-publish.yml]: A secret, to be more specific, e.g. `${{ secrets.DOCKER_USER }}`

**Comment by vsoch** [comment]
Okay `DOCKER_USER` and `DOCKER_PASSWORD` are added to secrets, when we can change the envar for the username to come from there, it could be ready to try out!

**Pull request merged**